### ISSUE TITLE
Hotkey interception in the filtering does not allow to use browser's search on the page

### ISSUE THREAD

**Issue by Alex-302** [AdguardTeam/AdguardBrowserExtension#2155]
### Issue Details

- **AdGuard version:** 4.1.1
- **Browser and version:** Chrome 103
- **Operating system and version:** Windows 10
- **Filters you use in AdGuard:** Base, 2x AdGuard tracking, Social, Annoyances, Russian
- **Stealth Mode settings:** All, except cookies
- **Other browser extensions:** no

### Actual Behavior

Cannot use Ctrl-F to search a text on the filtering log page. We use it to search a specific text in the lines, that displayed by the search, and on details panel.

### Expected Behavior

If possible, open browser's search by double Ctrl-F(or suggest another solution).

**Comment by maximtop** [comment]
We can turn off this shortcut for the filtering log if users prefer to use native search.

**Comment by contribucious** [comment]
#### :bulb: Alternative idea

Just pop up a small black or white _(theme awareness)_ semi-transparent notification (_custom toast notification_ … or native one if supported/enabled by the user) at bottom right for a few seconds after the user presses `Ctrl+F` to inform him like so: "Use `F3` instead for native search."

Possibly display this notification only once per "session" (versus at each `Ctrl+F`).
&nbsp;
&nbsp;
___
###### `F3` is OK at least on Firefox and Edge browsers on Windows 10.
###### However, should work for any Chromium/Firefox based browsers. Just make sure of that in many various situations though.

**Comment by Alex-302** [comment]
> "Use `F3` instead for native search."

It is obvious:) I just forgot about that.

> We can turn off this shortcut for the filtering log

Ok.

**Comment by contribucious** [comment]
:relaxed:
&nbsp;

> > We can turn off this shortcut for the filtering log
> 
> Ok.

_Personally_, I find it convenient.
As an alternative, there are at least:
- my solution (i.e. `Ctrl+F` works as currently but the user is made aware of the alternative access to the native search);
- the use of `Ctrl+Shift+F` instead to trigger the AdGuard search`*`, leaving `Ctrl+F` untouched;
- … or let the user choose as an advanced setting (but this adds an additional setting).

###### `*` User informed of this keyboard shortcut via tooltip when hovering over the "magnifying glass" icon.

**Comment by contribucious** [comment]
#### :thought_balloon: Regarding my solution …

<details>
<summary>Click to read …</summary>

#### Example found:

&nbsp;&nbsp;&nbsp;&nbsp; _**React Toast notification with react-hot-toast**_
&nbsp;&nbsp;&nbsp;&nbsp; https://codepen.io/piyushpd139/pen/qBpjaMa

&nbsp;&nbsp;&nbsp;&nbsp; :one: `<Toaster  position="top-right" />` in **JS (Babel)** part is actually ideal. As … this is the position where the user expects to find his search field on Chromium-based browsers, except he will find an informative notification at this location instead, basically telling him during a few seconds to use `F3` instead if native search desired — without preventing the current AdGuard search action so. _Also, this position does not hide the log lines. But "bottom-right" remains a possibility as well._

&nbsp;&nbsp;&nbsp;&nbsp; :two: _Possibly_ use the `Emoji Icon` version, but with a beautiful round informative icon instead (like that but in better: :information_source:).

&nbsp;&nbsp;&nbsp;&nbsp; :three: _Possibly_ use a simple fade-in / fade-out effect instead.
&nbsp;

To be displayed within the `Filtering log` window only when the user presses `Ctrl+F`.
But ideally, should not be displayed again until the next closing/opening of this window — or even, until the next closing/opening of the browser.
</details>

**Comment by contribucious** [comment]
:x: **UPD:** My solution, while "OK" on Edge (and probably on any Chromium-based browsers), is problematic on Firefox.

<details>
<summary>Click to read …</summary>
&nbsp;

**Reason:** Because globally, if the user closes the native search and tries to press `F3` again, this will trigger the search for the next occurrence (so the behavior of this key is not totally identical to `Ctrl+F` first) BUT … where Edge will redisplay the search field, Firefox will only highlight the next occurrence(s) _without_ displaying the search field again while doing so (meaning … search field no longer displayable therefore …).

:heavy_check_mark: **Alternative to consider consequently _(if native search really desired)_:** Switching to the second alternative, i.e. intercepting only `Ctrl+Shift+F`, leaving `Ctrl+F` untouched (see details in the post where I list the possibilities above).
&nbsp;
___
###### :warning: In any case, if we leave access to a (proper like so) native search via `Ctrl+F`, we will have to correct a more general problem currently visible with `F3`: Both Edge and Firefox have their search highlights removed after 1-2 sec. (probably related to a non-stop auto-refresh of logs by the extension, _possibly_ easily fixable but …).
</details>

**Comment by contribucious** [comment]
@Alex-302 By the way …

<details>
<summary>Click to reveal …</summary>
&nbsp;

Not important at all, but it should be noted that in the `Expected Behavior` part of your issue, your sentence <code>If possible, open browser's search by double Ctrl-F<strong>(or suggest another solution)</strong>.</code> can be interpreted in two ways:
- _or suggest another solution_ … to the user if he presses `Ctrl+F` twice, like to use `F3` instead (in which case, my main suggestion is very similar to yours, except pressing `Ctrl+F` one time suffices for me to inform the user + I specify the method, i.e. via a toast notification — appearing only once maximum per "session");
- _or suggest another solution_ … directed to the people / devs who read your issue — in the sense of "if you have any idea …" / "if you have a better idea …".

:arrow_right: I must say that I understood the second possibility. :relaxed:
</details>

**Comment by Alex-302** [comment]
@contribucious 

> * _or suggest another solution_ … directed to the people / devs who read your issue

This)

**Comment by contribucious** [comment]
## :point_up_2: Update

##### I updated the list of [possible alternatives](https://github.com/AdguardTeam/AdguardBrowserExtension/issues/2155#issuecomment-1206664828).

###### Long, I grant you, but I hope much clearer for the team of developers as well. :thumbsup:<br>Think of it more as a blog post actually. :wink:

**The votes are open!**
_(@Alex-302 Any preference? :relaxed:)_

**Comment by maximtop** [comment]
Seems like `F3` isn't working on macos.
`cmd + f + f` - not sure if extension can retransmite cmd-f shortcut to the native system
3, 4 - does the issue happens only when preview window is open?

I prefer `cmd/ctrl + shift + f` for focusing search field. This option looks valid for me.

**Comment by Alex-302** [comment]
> ctrl + shift + f

This hotkey is inconvenient to press with one hand.

**Comment by contribucious** [comment]
> Seems like `F3` isn't working on macos.

So, one less option. I kinda suspected that when I saw [this](https://github.com/AdguardTeam/AdguardBrowserExtension/commit/d71b6b58f9991802d0bc43a975e43c87f600a701#diff-cb33122f08a5b8098088ac9f1839a8eef5959e484e4334f4aadb164198629b9fL46) later — the `isWindowsOs` part — (taken from [this commit](https://github.com/AdguardTeam/AdguardBrowserExtension/commit/76ba98e311778d06a416140d19060ce429cc4a2d)).

:arrow_right: A variant could _however_ be the opposite: Leave `Ctrl+F` natively and intercept `F3` for AdGuard search (if this is not problematic on macOS either). But people may want a "next result" hotkey, therefore not an ideal choice. Another free `F[1-12]` key perhaps? (Often used by various apps globally on the system though. Not ideal either, consequently.)
&nbsp;

> I prefer `cmd/ctrl + shift + f` for focusing search field. This option looks valid for me.

I like this option too. One that came to me among the quickest by the way and quite naturally.

:bulb: `Cmd+Shift+F` (for macOS users) seems less easy than `Ctrl+Shift+F` though. Maybe `Ctrl+Shift+F` for all? (A bit less natural / macOS-style for them though.)
:eyes: Note that, in all cases, some users (both PC and MAC) have the FN key below SHIFT, instead of CTRL. Not my case but indeed, less practical then.
&nbsp;
&nbsp;
___
#### (Not _directly_ related to the subject itself.)
> 3, 4 - does the issue happens only when preview window is open?

Not sure I fully understand but here is the problem in a nutshell (Steps to reproduce).
<details>
<summary>Reveal …</summary>

1. Open `Filtering Log` of `AG for Windows`;
1. Visit https://example.org;
2. Click on the corresponding HTML entry in it;
3. Click on the `General` tab (default one though) and press `Ctrl+F` → MAIN field is focused as expected. :heavy_check_mark:
4. Click on the `Network` tab and press `Ctrl+F` → MAIN field is focused as expected. :heavy_check_mark:
5. Click on the `Preview` tab and press `Ctrl+F` → MAIN field is focused. CODE field should be instead in my opinion. Like [this](https://user-images.githubusercontent.com/4764956/183290233-2b511eb6-d50f-4778-878a-b8a03d336de9.png), to reuse one of my screenshots. _Without having to click in the code first._ However, I understand the current logic (i.e. if changed to my idea, how to easily access the main field again _if_ you stay in this tab from then on?)

:arrow_right: Anyway, if alternative :four: is implemented (i.e. a new code search field is added in the extension too), then ideally it should be standardized on both (app(s) and extension). Meaning … either clicking in the code itself required beforehand to search some code via `Ctrl+F` for all "products" or for none.

:arrow_right: And if alternative :three: is implemented (i.e. browser's native search allowed only in one place: the `Preview` section), since the browser's native search starts its search directly without worrying about having clicked in the code beforehand, we should logically do the same with `AdGuard for Windows` with its dedicated field (i.e. as soon as the code is physically visible to our eyes, we should be able to `Ctrl+F` and type our search). But as I said, I understand the current logic of `AG for Windows` though.
</details>

**Comment by Alex-302** [comment]
> open browser's search by double Ctrl-F

For example, Slack on the first `Shift-Esc` suggests to clear unread messages, but if you press the hotkey twice - it opens browser's task manager(in Chrome).

**Comment by contribucious** [comment]
> Seems like `F3` isn't working on macos.

:bulb: After research, the non OS-specific version of this keyboard shortcut (used by all major browsers for `Find next`) is:
**`Ctrl+G`** (Windows/Linux) / **`Command-G`** (macOS) &nbsp;—&nbsp; _Vote post [adequately updated](https://github.com/AdguardTeam/AdguardBrowserExtension/issues/2155#issuecomment-1206664828)._

<details>
<summary>Still not a viable solution though …</summary>
&nbsp;

**Because** of the same problematic behavior as `F3` on Firefox, described [here](https://github.com/AdguardTeam/AdguardBrowserExtension/issues/2155#issuecomment-1206809790).

&nbsp;&nbsp;&nbsp; :arrow_right_hook: **Can be used as a _standard_ workaround in the meantime however.**

&nbsp;&nbsp;&nbsp; :arrow_right_hook: **Bonus FF fix:** If a Firefox user really needs to do a second search, in the meantime, he can still however UNFOCUS the AdGuard main search field and type `/` (to let the _basic_ `Quick Find` bar appear).
&nbsp;
</details>
___
<details>
<summary>:mag: Bonus note …</summary>

#### About the Win/Linux specific `F3` (macOS [too](https://support.mozilla.org/en-US/kb/keyboard-shortcuts-perform-firefox-tasks-quickly#firefox:mac:fx103) for Firefox):
While described as `Find next` action on [Firefox](https://support.mozilla.org/en-US/kb/keyboard-shortcuts-perform-firefox-tasks-quickly#firefox:win10:fx103) _("Find Again" using their words)_ and has been for a while on [Chrome too](https://apple.stackexchange.com/a/229487), on the latter, this hotkey is [now listed](https://support.google.com/chrome/answer/157179?hl=en#zippy=%2Cgoogle-chrome-feature-shortcuts) as a alternative for `Find` action (i.e. `Ctrl+F`) despite its _strictly_ equivalent behavior to `Find next` (i.e. `Ctrl+G`). Probably to clearly signal to Chrome users that on their browser a new search is _always possible_ via this hotkey (putting aside the — yet main — fact that multiple presses on it _will_ search for the next occurrence).
</details>

**Comment by contribucious** [comment]
###### _(Due to the workaround above, there is no rush.)_

> I prefer `cmd/ctrl + shift + f` for focusing search field. This option looks valid for me.

**I note an advantage to this light/quick solution:**
you can keep all native stuff with `Ctrl + [key]` and set AdGuard specific stuff with `Ctrl + Shift + [key]`.

**For example** _(flexible solution that can even be extended in the future)_:

<details>
<summary>:camera: Reveal …</summary>

&nbsp;
![lF05TXfKYx](https://user-images.githubusercontent.com/4764956/187268870-ad5f4f61-ca86-4997-afd5-320aa65d0912.png)
:bulb: _Just use a tooltip on the respective icons to inform the user about these keyboard shortcuts._

##### Native:
`Ctrl + F` **→** search within the "filtering log page"
`Ctrl + R` **→** reload the "filtering log page"

##### AdGuard specific stuff:
`Ctrl + Shift + F` **→** search within requests
`Ctrl + Shift + R` **→** reload the associated tab (i.e. `example.org` here) :new:

###### _(`Ctrl` or … `Cmd` on macOS.)_
</details>

**Comment by contribucious** [comment]
<details>
<summary>:1st_place_medal: About this flexibility …</summary>
&nbsp;

Also, Chrome (and [Edge, as detailed here](https://techdows.com/2022/02/microsoft-edge-tab-search.html) — in the stable channel now) seems to have chosen to use `Ctrl + Shift + [key]` when it comes to _special_ tab management. For example, search for a tab within your list of tabs with `Ctrl + Shift + A`.

Because `Filtering log` is a tabless popup window, those specific native hotkeys have no effect there. And that's perfect in our situation! Thus, AdGuard can for example search its own tab list similarly using `Ctrl + Shift + A` too **(compare this [AdGuard screenshot](https://user-images.githubusercontent.com/4764956/187280666-27bc3acc-8f8f-4b86-ad5b-80c073239e8e.png) and this [Edge screenshot](https://user-images.githubusercontent.com/4764956/187283530-3eb08038-ae6d-4fba-9c8c-1c54d17577f1.png))**.

After testing under Edge and Firefox, the interception works well. 🙂
Note just that under the latter, it hijacks access to "about:addons" though, but that's okay (because only in this popup). :thumbsup:
&nbsp;
</details>

<details>
<summary>:gift: Possible bonus …</summary>

##### By the way, if implemented, in the continuity and if you agree, I'll create another issue (`Feature Request`) to have _classic_ tab management replicated in `Filtering log` too:
- `Ctrl + Tab` / `Ctrl + Shift + Tab`;
- `Ctrl + [1-9]`.

:arrow_right_hook: _AdGuard managing this in relation to its own list of tabs._
</details>
